src/runtime_src/core/common/memaccess.cpp

```diff
@@ -0,0 +1,200 @@
+// Sanity check the user's Start Address and Size against the mem topology.
+// If the start address is 0 (ie. unspecified by user) change it to the first available address.
+// If the size is 0 (ie. unspecified by user) change it to the maximum available size.
+// Fill the vector with the available banks.
+// Set the iterator to the bank containing the start address.
+static void
+validate_address_and_size(xrt_core::device* device, uint64_t& start_addr, uint64_t& size,
+            std::vector<mem_bank_t>& vec_banks, std::vector<mem_bank_t>::iterator& start_bank) 
+{
+  // Validate the memory banks
+  vec_banks = get_ddr_banks(device);
+  // Find the first memory bank with valid size since vec_banks is sorted
+  auto valid_bank =  std::find_if(vec_banks.begin(), vec_banks.end(),
+              [](const mem_bank_t item) {
+                return item.m_size;
+              });
+
+  if (valid_bank == vec_banks.end())
+    throw xrt_core::error(std::errc::operation_canceled, "ERROR: Couldn't find valid memory banks");
+
+  // Validate the start address
+  // If given start address is 0 choose a start address of the first available memory bank
+  // Update reference
+  start_addr = (start_addr == 0) ? valid_bank->m_base_address : start_addr;
+
+  //Sanity check start address
+  start_bank = std::find_if(vec_banks.begin(), vec_banks.end(),
+              [=](const mem_bank_t& item) {
+                return ((start_addr >= item.m_base_address) && (start_addr < (item.m_base_address+item.m_size)));
+              });
+
+  if (start_bank == vec_banks.end())
+    throw xrt_core::error(std::errc::operation_canceled,  boost::str(boost::format("Start address 0x%x is not valid") % start_addr));
+
+  // Validate the amount of accessable memory
+  uint64_t available_size = std::accumulate(start_bank, vec_banks.end(), (uint64_t)0,
+          [](uint64_t result, const mem_bank_t& obj) {
+            return result + obj.m_size;
+          }) ;
+
+  available_size -= start_addr - start_bank->m_base_address;
+  if (size > available_size) 
+    throw xrt_core::error(std::errc::operation_canceled, boost::str(boost::format("Cannot access %d bytes of memory from start address 0x%x\n") % size % start_addr));
+
+  // If no size is supplied the end address is the ending address of the last available memory bank
+  // Update the input reference to reflect the change
+  size = (size == 0) ? available_size : size;
+}
+
+enum class operation_type {
+  read,
+  write
+};
+
+// Ensure safe access into a device's memory banks based on memory bank boundary and if the bank is in use
+static void 
+perform_memory_action(xrt_core::device* device, xrt_core::aligned_ptr_type& buf, uint64_t start_addr, uint64_t size, operation_type action)
+{
+  std::vector<mem_bank_t> vec_banks;
+  std::vector<mem_bank_t>::iterator start_bank;
+
+  // Sanity check the address and size against the mem topology
+  validate_address_and_size(device, start_addr, size, vec_banks, start_bank);
```

Comment:
Horrible interface of validate function is reflected in these lines.  This is so procedural, it hurts to look at it.
Again, I know it is not your code, but do you want to clean it up?

Reply:
I split the function into three pieces: One for each parameter it used to return. Looks much better! I opted against putting everything into the operation function as I like breaking up functions to make it a little easier to see what is going on. Let me know if you'd rather have the functions within the operation function.